sklearn/utils/arrayfuncs.pyx

```diff
@@ -24,6 +32,48 @@ def min_pos(const floating[:] X):
     return min_val
 
 
+def _all_with_any_reduction(real_numeric[:, :] array, real_numeric value=0, int axis=0):
+    """Check that all values are equal to `value` along a specific axis.
+
+    It is equivalent to `np.any(np.all(X == value, axis=axis))`, but it avoids to
+    materialize the temporary boolean matrices in memory.
+
+    Parameters
+    ----------
+    array: array-like
+        The array to be checked.
+    value: short, int, long, float, or double, default=0
+        The value to use for the comparison.
+    axis: int, default=0
+        The axis along which to perform the `all` operation.
```

Comment:
Given we only need `axis=1` in `KNeighborsClassifier`, can we restrict this function to: `_all_with_any_reduction_axis_1`?

Reply:
Sure, I will make this change for now. Indeed I cannot easily come up with a scenario that we want to check "some feature of all samples is equal to a value". But @glemaitre if you are against this change I can revert.